napari/utils/_magicgui.py

```diff
@@ -0,0 +1,136 @@
+of those custom classes, magicgui will know what to do with it.
+
+"""
+from typing import Tuple, Type, Any
+from ..layers import Layer
+
+try:
+    from magicgui import register_type as _magictype
+except ImportError:
+
+    def _magictype(*args, **kwargs):
+        pass
+
+
+def register_types_with_magicgui():
+    """Register napari types with magicgui.
+
+    Parameter Annotations -> Widgets:
+        napari.layers.Layer, will be rendered as a ComboBox.
+            if a parameter is annotated as a subclass Layer type, then the
+            combobox options will be limited to that layer type.
+
+    Return Annotations -> Widgets:
+        napari.layers.Layer will add a new layer to the Viewer.
+            if a return is annotated as a subclass of Layer, then the
+            corresponding layer type will be added.
+            see `show_layer_result` for detail
+
+    """
+    _magictype(Layer, choices=get_layers, return_callback=show_layer_result)
+
+
+def get_layers(gui, layer_type: Type[Layer]) -> Tuple[Layer, ...]:
+    """Retrieve layers of type `layer_type`, from the Viewer the gui is in.
+
+    Parameters
+    ----------
+    gui : MagicGui or QWidget
+        The instantiated MagicGui widget.  May or may not be docked in a
+        dock widget.
+    layer_type : type
+        This is the exact type used in the type hint of the user's
+        function. It may be a subclass of napari.layers.Layer
+
+    Returns
+    -------
+    tuple
+        Tuple of layers of type ``layer_type``
+
+    Example
+    -------
+    This allows the user to do this, and get a dropdown box in their GUI
+    that shows the available image layers.
+
+    >>> @magicgui
+    ... def get_layer_mean(layer: napari.layers.Image) -> float:
+    ...     return layer.data.mean()
+
+    """
+    try:
+        # look for the parent Viewer based on where the magicgui is docked.
+        # if the magicgui widget does not have a parent, it is unattached
+        # to any viewers, and therefore we cannot return a list of layers
+        viewer = gui.parent().qt_viewer.viewer
+        return tuple(l for l in viewer.layers if isinstance(l, layer_type))
+    except AttributeError:
+        return ()
+
+
+def show_layer_result(gui, result: Any, return_type: Type[Layer]) -> None:
+    """Show a magicgui result in the viewer.
+
+    Parameters
+    ----------
+    gui : MagicGui or QWidget
+        The instantiated MagicGui widget.  May or may not be docked in a
+        dock widget.
+    result : Any
+        The result of the function call.
+    return_type : type
+        The return annotation that was used in the decorated function.
+    """
+    if result is None:
+        return
+
+    try:
+        viewer = gui.parent().qt_viewer.viewer
+    except AttributeError:
+        return
+
+    # if they have annotated the return type as layers.Layer,
+    # the function MUST return a list of LayerData tuples
```

Comment:
(a) can you define what a LayerData tuple is in this comment?

(b) can you clarify "if they have annotated the return type as a base layer (layers.Layer), NOT a subclass of it, then ...". I originally read this as including subclasses and was highly confused. =)